presto-main/src/test/java/com/facebook/presto/TestClientRequestFilterPlugin.java

```diff
@@ -0,0 +1,298 @@
+                    .allMatch(headerName -> request.getHeader(headerName) != null);
+
+            if (!headersPresent) {
+                Optional<Map<String, String>> extraHeaderValueMap = requestFilter.getExtraHeaders(testPrincipal);
+
+                extraHeaderValueMap.ifPresent(map -> {
+                    for (Map.Entry<String, String> extraHeaderEntry : map.entrySet()) {
+                        String headerKey = extraHeaderEntry.getKey();
+                        if (headersBlockList.contains(headerKey)) {
+                            throw new RuntimeException("Modification attempt detected: The header " + headerKey + " is present in the blocked headers list.");
+                        }
+                        if (globallyAddedHeaders.contains(headerKey)) {
+                            throw new RuntimeException("Header conflict detected: " + headerKey + " already added by another filter.");
+                        }
+
+                        if (request.getHeader(headerKey) == null && requestFilter.getHeaderNames().contains(headerKey)) {
+                            extraHeadersMap.putIfAbsent(headerKey, extraHeaderEntry.getValue());
+                            globallyAddedHeaders.add(headerKey);
+                        }
+                    }
+                });
+            }
+        }
+        request.setHeaders(extraHeadersMap);
+    }
+
+    private List<ClientRequestFilter> getClientRequestFilters()
+    {
+        List<ClientRequestFilter> requestFilters = new ArrayList<>();
+        ClientRequestFilter customModifier = new ClientRequestFilter()
+        {
+            @Override
+            public List<String> getHeaderNames()
+            {
+                return Collections.singletonList("X-Custom-Header");
+            }
+            @Override
+            public <T> Optional<Map<String, String>> getExtraHeaders(T additionalInfo)
+            {
+                Map<String, String> headers = new HashMap<>();
+                headers.put("X-Custom-Header", "CustomValue_1");
+                return Optional.of(headers);
+            }
+        };
+
+        ClientRequestFilter customModifierConflict = new ClientRequestFilter()
+        {
+            @Override
+            public List<String> getHeaderNames()
+            {
+                return Collections.singletonList("X-Custom-Header");
+            }
+            @Override
+            public <T> Optional<Map<String, String>> getExtraHeaders(T additionalInfo)
+            {
+                Map<String, String> headers = new HashMap<>();
+                headers.put("X-Custom-Header", "CustomValue_2");
+                return Optional.of(headers);
+            }
+        };
+
+        requestFilters.add(customModifier);
+        requestFilters.add(customModifierConflict);
+        return requestFilters;
+    }
+
+    static class ConcreteHttpServletRequest
+            extends MockHttpServletRequest
```

Comment:
Why not modify `MockHttpServletRequest` directly?

Reply:
In ConcreteHttpServletRequest, I am storing additional headers in the customHeaders map and overriding the getHeaders method to manage them. This functionality isn't natively provided by MockHttpServletRequest, so extending it makes sense.

Reply:
Just add a new constructor to `MockHttpServletRequest` that overrides the headers.